app/Main.hs

```diff
@@ -1,16 +1,26 @@
-{-# LANGUAGE FlexibleContexts #-}
 module Main where
 
 import           Control.Monad.State.Lazy
 import           Data.Char
+import           Data.Maybe
 import           System.Environment
 
-initialCellState = replicate 30000 (0 :: Int)
-
 type CellState = [Int]
 
--- MoveState :: Int (Position) -> Parsed Commands -> CellState
-type MoveState = (Int, [Command], CellState)
+data MoveState
```

Comment:
Better to use a record instead of a large tuple. You ended up writing a comment to remember which field was which. This way you don't have to remember.